bika/lims/browser/fields/uidreferencefield.py

```diff
@@ -207,5 +179,121 @@ def set(self, context, value, **kwargs):
+    :param value: A UID.
+    :type value: string
+    :return: True if the value is a UID and exists as an entry in uid_catalog.
+    :rtype: bool
+    """
+    uc = getToolByName(context, 'uid_catalog')
+    brains = uc(UID=value)
+    return brains and True or False
+
+
+def _get_object(context, value):
+    """Resolve a UID to an object.
+
+    :param context: context is the object containing the field's schema.
+    :type context: BaseContent
+    :param value: A UID.
+    :type value: string
+    :return: Returns a Content object.
+    :rtype: BaseContent
+    """
+
+    if value:
+        if is_at_content(value) or is_dexterity_content(value):
+            return value
+        else:
+            uc = getToolByName(context, 'uid_catalog')
+            brains = uc(UID=value)
+            if brains:
+                return brains[0].getObject()
+
+
+def get_storage(context):
+    annotation = IAnnotations(context)
+    if annotation.get(BACKREFS_STORAGE) is None:
+        annotation[BACKREFS_STORAGE] = {}
+    return annotation[BACKREFS_STORAGE]
+
+
+@mutually_exclusive_parameters('as_uids', 'as_objects')
+def get_backreferences(context, relationship=None,
+                       as_uids=None, as_objects=None):
+    """Return all objects which use a UIDReferenceField to reference context.
+
+    :param context: The object which is the target of references.
+    :param relationship: The relationship name of the UIDReferenceField.
+    :param as_uids: when a relationship is specified, requests only UIDs.
+    :param as_objects: when a relationship is specified, requests full objects.
+
+    This function can be called with or without specifying a relationship.
+
+    - If a relationship is provided, the return value will be a list of items
+      which reference the context using the provided relationship.
+
+      If relationship is provided, then you can provide an extra argument
+      to specify how the back references should be returned: as UIDs,
+      full objects, or brains (the default).
+
+    - If the relationship is not provided, then the entire set of
+      backreferences to the context object is returned (by reference) as a
+      dictionary.  This value can then be modified in-place, to edit the stored
+      backreferences.
+
+      If relationship is not provided, the returned backreferences will
+      always be returned as UIDs, and it is an error to specify another
+      return type.
+    """
+
+    instance = context.aq_base
+    backrefs = get_storage(instance)
+
+    if relationship:
+        backrefs = backrefs.get(relationship, [])
+        if not backrefs:
+            return backrefs
+        # return only UIDs
+        if as_uids:
+            return backrefs
+        # return full objects
+        elif as_objects:
+            return [_get_object(context, b) for b in backrefs]
+        # Return brains from first catalog found.
+        cat = _get_catalog_for_uid(backrefs[0])
+        return cat(UID=backrefs)
```

Comment:
This logic seems to me confusing. Either we should return UIDs or objects depending on the parameters the caller of this function provided.

Better refactor this function as follows:

```
instance = context.aq_base
storage = get_storage(instance)
if relationship is None:
    assert not as_objects, "You cannot use as_objects with no relationship"
    return storage
backrefs = backrefs.get(relationship, None)
if backrefs in None:
    return []
if as_objects:
    return [_get_object(context, b) for b in backrefs]
return backrefs
```